### PR TITLE
feat: Bulk API Implementations

### PR DESCRIPTION
**Pull request by akrock**
Adding Bulk APIs for DataStores.

TODO:  Run Unit Tests
* Add general bulk tests
* Add cosmos specific bulk tests?

### CODE REVIEW

File: DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs [hunk 1]
```diff
@@ -47,11 +51,7 @@ public class CosmosDbDataStore<TKey, TValue> :
 
         readonly Lazy<Container> _container;
         
-        readonly string _connectionString;
-        
-        readonly string _databaseId;
-        
-        readonly string _containerId;
```
Comment (halforbit) [inline DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs:54]: These should not get removed. While they arent used by the implementation, these fields are used by the builder integration tests to verify that the proper values are propagated. See `Halforbit.DataStores.IntegrationTests.FluentBuilderTests.CosmosDb()`.

File: DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs [hunk 2]
```diff
@@ -610,7 +684,136 @@ async Task ObserveBeforePut(TKey key, TValue value)
...
+                foreach (var batch in grp.Batch(BulkApiMaxOpsPerRequest))
+                {
+                    var useBulk = batch.Count() > BulkApiMinOpsPerPartition;
+                    if (useBulk)
+                    {
+                        bulkTasks.Add(OperationWrapper(semaphore, batch,
+                            async b =>
+                            {
+                                var results = b.Select(async i => new KeyValuePair<TKey, TOut>(i.Key,
+                                    await operation(_bulkContainer.Value, i.PartitionKey, i.DocumentId, i.Key, i.Value)));
```
Comment (halforbit) [inline DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs:724]: we probably need to put a `.ConfigureAwait(false)` on _all_ usages of `await`, both here and elsewhere in these changes.

File: Halforbit.DataStores/Extensions/EnumerableExtensions.cs [hunk 3]
```diff
@@ -91,5 +93,56 @@ IEnumerable<TResult> Batch(int sz)
                 }
             }
         }
+    
+        
+        internal static async Task<IReadOnlyList<TOut>> ParallelSelectAsync<TIn, TOut>(this IEnumerable<TIn> inputs, Func<TIn, Task<TOut>> mutator, CancellationTokenSource cancellationTokenSource = null, int maxPending = int.MinValue)
```
Comment (halforbit) [inline Halforbit.DataStores/Extensions/EnumerableExtensions.cs:98]: Is there a reason these can't be simply `SelectAsync` and `ForEachAsync`?

File: DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs (hunk 1 above)
Comment (akrock) [reply DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs]: Fixed in 6c69825f2eebd4a96da9e4b233094040d38fdfb9

File: Halforbit.DataStores/Extensions/EnumerableExtensions.cs (hunk 3 above)
Comment (akrock) [reply Halforbit.DataStores/Extensions/EnumerableExtensions.cs]: Nope I just straight stole them from our PC implementations. Renamed in deac7feedd35f2add9219874e4cf76b76ce4e3ef

File: DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs (hunk 2 above)
Comment (akrock) [reply DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs:736]: I tried to be cognizant of that, apparently I missed a bunch on the changes in this file.

Comment (akrock) [reply DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs]: Seemed like a large scale problem throughout the code base. So I went ahead and addressed them all.

5e11e77d6c22b1bc6bfd8dcd266f86e06476bfd7

**Comment by akrock** [comment]
I added new tests now to stress the builk implementations now in: 8e59a1fbe20f06e0629d927ca547896fcc4bd441

I ran the partition tests for cosmos & the blob storage tests, everything looks good but you might want to run the full integration test suite.

**Pull request merged**